cpp/src/arrow/compare.cc

```diff
@@ -984,7 +984,8 @@ bool ArrayRangeEquals(const Array& left, const Array& right, int64_t left_start_
   bool are_equal;
   if (&left == &right) {
     are_equal = true;
-  } else if (left.type_id() != right.type_id()) {
+  } else if (left.type_id() != right.type_id() ||
+             !TypeEquals(*left.type(), *right.type(), false /* check_metadata */)) {
```

Comment:
Can you add a couple of tests with some parametric types (e.g. TIMESTAMP with different time units)?

Reply:
Thanks wes, fixed.